eth/stagedsync/stage_blockhashes.go

```diff
@@ -0,0 +1,59 @@
+package stagedsync
+
+import (
+	"bytes"
+
+	"github.com/ledgerwatch/turbo-geth/common"
+	"github.com/ledgerwatch/turbo-geth/common/dbutils"
+	"github.com/ledgerwatch/turbo-geth/core/rawdb"
+	"github.com/ledgerwatch/turbo-geth/ethdb"
+	"github.com/ledgerwatch/turbo-geth/log"
+)
+
+func SpawnBlockHashStage(s *StageState, stateDB ethdb.Database, quit <-chan struct{}) error {
+	progress := s.BlockNumber
+	batch := stateDB.NewBatch()
+	firstHash := s.StageData
+	headHash := rawdb.ReadHeadHeaderHash(stateDB).Bytes()
+	if err := stateDB.Walk(dbutils.HeaderPrefix, dbutils.HeaderKey(progress, common.BytesToHash(firstHash)), 0, func(k []byte, _ []byte) (bool, error) {
+		if err := common.Stopped(quit); err != nil {
+			return true, err
+		}
+
+		if len(k) != 40 {
+			return true, nil
+		}
+		if err := batch.Put(dbutils.HeaderNumberPrefix, k[8:], k[:8]); err != nil {
```

Comment:
k - will valid only during transaction, but you do .Commit outside, then need copy it manually

Reply:
Good catch